lib/TPP/ToBlockLayoutAndBack.cpp

```diff
@@ -507,27 +510,50 @@ mlir::linalgx::packVNNIMatmulOp(RewriterBase &rewriter,
       toPackLayout_VNNI(rewriter, loc, matmulOp.getInputs()[1], tilesOnB);
   MLIRContext *ctx = matmulOp.getContext();
   AffineExpr p1, p2, r1, p3, p4, r2, r3;
-  bindDims(ctx, p1, p2, r1, p3, p4, r2, r3);
   SmallVector<Value> packedInputs = {matmulOp.getInputs()[0], packedMatrixB};
-  AffineMap mapA =
-      AffineMap::get(/*dims=*/7, /*symbols=*/0, {p1, r1, p3, r2}, ctx);
-  AffineMap mapB =
-      AffineMap::get(/*dims=*/7, /*symbols=*/0,
-                     {p2, r1, r2.floorDiv(*blockingFactor), p4, r3}, ctx);
-  AffineMap mapC =
-      AffineMap::get(/*dims=*/7, /*symbols=*/0, {p1, p2, p3, p4}, ctx);
+  int64_t dims;
+  AffineMap mapA, mapB, mapC;
   Value matrixC = matmulOp.getOutputs()[0];
-  linalg::GenericOp replacementOp = rewriter.create<linalg::GenericOp>(
-      loc, matrixC.getType(), packedInputs, ValueRange{matrixC},
-      ArrayRef<AffineMap>{mapA, mapB, mapC},
-      ArrayRef<mlir::utils::IteratorType>{mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::reduction,
-                                          mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::reduction,
-                                          mlir::utils::IteratorType::reduction},
-      /*doc=*/"", /*libraryCall=*/"");
+  linalg::GenericOp replacementOp;
+  if (matmulOp.getInputs()[1].getType().cast<ShapedType>().getRank() == 4) {
```

Comment:
why do we have to differentiate here?

Reply:
The maps and iterators to linalg.generic differ for operands of different shapes (i.e., when shape is 4d, generic is 7 dimensional and maps are as per the code, as compared to 3d shape obtained when using tiling factor of 1,1).

Reply:
Now I am worried. If we start adding a case for each combination, this does not scale.

Reply:
If tiling is a 1,1 it tiles along parallel dimensions and indeed this seems the case looking at the affine map, I don't think we need to differentiate, we should look only at the last [r, r, p, p, r] dimensions.

Reply:
Added a todo for now.